openedx_tagging/core/tagging/rules.py

```diff
@@ -53,17 +64,48 @@ def can_change_tag(user: UserType, tag: Tag | None = None) -> bool:
 
 
 @rules.predicate
-def can_change_object_tag(user: UserType, object_tag: ObjectTag | None = None) -> bool:
+def can_change_object_tag_taxonomy(user: UserType, taxonomy: Taxonomy) -> bool:
     """
-    Taxonomy admins can create or modify object tags on enabled taxonomies.
+    Everyone can create or modify object tags on enabled taxonomies.
+    Only taxonomy admins can create or modify object tags on disabled taxonomies.
+
     """
-    taxonomy = (
-        object_tag.taxonomy.cast() if (object_tag and object_tag.taxonomy) else None
-    )
-    object_tag = taxonomy.object_tag_class.cast(object_tag) if taxonomy else object_tag
-    return is_taxonomy_admin(user) and (
-        not object_tag or not taxonomy or (taxonomy and taxonomy.enabled)
-    )
+    taxonomy = taxonomy.cast()
+
+    return bool(taxonomy and taxonomy.enabled) or is_taxonomy_admin(user)
+
+
+@rules.predicate
+def can_change_object_tag_objectid(_user: UserType, _object_id: str) -> bool:
+    """
+    Nobody can create or modify object tags without checking the permission for the tagged object.
+
+    This rule should be defined in other apps for proper permission checking.
```

Comment:
Nice!

Reply:
We have some options here:
1. return True for this app (I think it is ok to tag everything here: we are tagging `str`)
2. return False and change the tests accordingly (everything will return 403 and will barely cover the implemented code in the rest_api)
3. return False and change the tests accordingly (everything will return 403) AND duplicate the tests patching this permission (forcing True) to actually test the workflow. 

In the actual context, `1` is the faster/easier path. For 2 and 3, I must fix the `test_rules.py` and `test_views.py`. Let me know what you think!

Reply:
In edx-platform, we'll be adding a rule that defines how "content objects" are tagged.

So in the tests, can't we add a new rule just for tests that defines how "Foo objects" are tagged? And then confirms that the rule is enforced by trying to tag some Foo objects?

In other words, the tests should work in a similar way to how the platform code is going to work. I would prefer no patching and I want to see the actual functionality being tested so I don't like 2 or 3.

Please review https://github.com/openedx/modular-learning/issues/98 , I created this ticket for next sprint to clearly define the missed requirements here around permissions. If you think these tests will take a while, we can just return True for now and make sure the permissions are properly tested when we implement that ticket but I'd prefer to make sure that the base functionality in openedx-learning is implemented correctly now.

Reply:
> So in the tests, can't we add a new rule just for tests that defines how "Foo objects" are tagged? And then confirms that the rule is enforced by trying to tag some Foo objects?

Awesome idea @bradenmacdonald!

Reply:
I would like to add some tests to check if we can't tag anything but "abc". Just give me a minute before merge!

Reply:
Thanks @rpenido , that's great. No worries, let me know when you want it merged.

Reply:
Phew! It had a copy/paste error. Tested and fixed 85b320a8aad20dab06dd931d26e20a20279ca790

I think we are good to go now @bradenmacdonald!

Reply:
Looks like a build error. No rush, just let me know when it's green and ready :)

Reply:
> I think we are good to go now @bradenmacdonald!

Too soon.
```
openedx_tagging/core/tagging/rules.py:92: error: Argument 2 to "has_perm" of "PermissionsMixin" has incompatible type "str"; expected "Optional[Union[Model, AnonymousUser]]"  [arg-type]
openedx_tagging/core/tagging/rules.py:92: error: Argument 2 to "has_perm" of "AnonymousUser" has incompatible type "str"; expected "Optional[Union[Model, AnonymousUser]]"  [arg-type]
```

Related:
- https://github.com/typeddjango/django-stubs/issues/231 (fix: https://github.com/typeddjango/django-stubs/pull/232)

But I think the fix is not enought..

Reply:
@rpenido According to [the django docs for `has_perm`](https://docs.djangoproject.com/en/4.2/ref/contrib/auth/#django.contrib.auth.models.User.has_perm) it seems like the second object is usually an object, but here we're passing in a string (the object ID). So I think the type check is mostly correct. BUT here we don't have objects, we have object IDs, and the API doesn't actually care how you define the permissions checks. I think defining them to use the object_id is quite reasonable. So just add a comment explaining that and put `#type: ignore[arg-type]` at the end of the line to suppress the warning.

Reply:
Thank you @bradenmacdonald! We are good now!²